src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+namespace Pim\Component\Catalog\Completeness;
+
+use Akeneo\Component\StorageUtils\Cursor\CursorInterface;
+use Akeneo\Component\StorageUtils\Remover\BulkRemoverInterface;
+use Akeneo\Component\StorageUtils\Saver\BulkSaverInterface;
+use Akeneo\Component\StorageUtils\Saver\SaverInterface;
+use Pim\Component\Catalog\Model\ChannelInterface;
+use Pim\Component\Catalog\Model\FamilyInterface;
+use Pim\Component\Catalog\Model\LocaleInterface;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderFactoryInterface;
+use Pim\Component\Catalog\Query\ProductQueryBuilderInterface;
+
+/**
+ * Simple object version of the completeness generator.
+ *
+ * @author    Julien Janvier (j.janvier@akeneo.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class CompletenessGenerator implements CompletenessGeneratorInterface
+{
+    const BULK_SIZE = 50;
+
+    /** @var ProductQueryBuilderFactoryInterface */
+    protected $pqbFactory;
+
+    /** @var CompletenessCalculatorInterface */
+    protected $completenessCalculator;
+
+    /** @var SaverInterface */
+    protected $productSaver;
+
+    /** @var BulkSaverInterface */
+    protected $productBulkSaver;
+
+    /** @var BulkRemoverInterface */
+    private $completenessRemover;
+
+    /**
+     * @param ProductQueryBuilderFactoryInterface $pqbFactory
+     * @param CompletenessCalculatorInterface     $completenessCalculator
+     * @param SaverInterface                      $productSaver
+     * @param BulkSaverInterface                  $productBulkSaver
+     * @param BulkRemoverInterface                $completenessRemover
+     */
+    public function __construct(
+        ProductQueryBuilderFactoryInterface $pqbFactory,
+        CompletenessCalculatorInterface $completenessCalculator,
+        SaverInterface $productSaver,
+        BulkSaverInterface $productBulkSaver,
+        BulkRemoverInterface $completenessRemover
+    ) {
+        $this->pqbFactory = $pqbFactory;
+        $this->completenessCalculator = $completenessCalculator;
+        $this->productSaver = $productSaver;
+        $this->productBulkSaver = $productBulkSaver;
+        $this->completenessRemover = $completenessRemover;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateMissingForProduct(ProductInterface $product)
+    {
+        $this->completenessCalculator->calculate($product);
+        $this->productSaver->save($product);
```

Comment:
Shouldn't you instead move the logic from the ProductSaver to here, so the ProductSaver use just the completeness manager like before?